### PR TITLE
Reduce memory usage in queries with `ORDER BY primary_key

### LINKED ISSUE #16250: [20.11] optimize_read_in_order works only in reverse direction.
**Describe the bug**
Optimize read in order doesn't work in specific conditions.

**How to reproduce**
Clickhouse server 20.11.1.4941

```
CREATE TABLE test_asof(key UInt32, ts DateTime) ENGINE=MergeTree PARTITION BY toYYYYMM(ts) ORDER BY (key,ts);
INSERT INTO test_asof SELECT number % 100000 as key, now() - toIntervalMonth(intDiv(number, 100000)) as ts FROM numbers(1000000);
set max_threads=1;

SELECT *
FROM test_asof
ORDER BY
    key DESC,
    ts DESC
LIMIT 1
FORMAT Null

Query id: d8a698d1-567b-4c01-956b-440fd7b5e70e

Ok.

0 rows in set. Elapsed: 0.008 sec. Processed 262.72 thousand rows, 2.10 MB (34.62 million rows/s., 276.97 MB/s.)

SELECT *
FROM test_asof
ORDER BY
    key DESC,
    ts ASC
LIMIT 1
FORMAT Null

Query id: 98dfb13c-716b-42dd-b4ee-5778f48d4690

Ok.

0 rows in set. Elapsed: 0.026 sec. Processed 590.40 thousand rows, 4.72 MB (22.73 million rows/s., 181.85 MB/s.)

SELECT *
FROM test_asof
ORDER BY
    key ASC,
    ts ASC
LIMIT 1
FORMAT Null

Query id: 1843cc1b-4a72-4820-9385-48cbaea1ed12

Ok.

0 rows in set. Elapsed: 0.012 sec. Processed 1.00 million rows, 8.00 MB (84.13 million rows/s., 673.03 MB/s.)
```

**Expected behavior**
optimize_read_in_order works in both directions.

### LINKED ISSUE #20553: Select * memory usage optimization
I use ClickHouse 20.12.3.3 to store normalized logs. A table consists of 160 columns, primary key and order by key is Timestamp (int64).

Often I need to find some logs extracting all table columns. When I use the following query to find logs in 1 month period (approx. 2 TB of compressed data) on a single node - it consumes 20+ GB of memory:

SELECT * FROM events WHERE Column1 = 'value1' AND Column2 = 'value2' AND Timestamp > [START] AND Timestamp < [END] ORDER BY Timestamp DESC LIMIT **250**;

The wider search period is, the more RAM is consumed. The less columns appear in SELECT, the less memory consumed.

Folks from CH telegram group told me that part of this memory is allocated for column buffers (1MB for each column that appears in SELECT) by each thread. Let's say query is executed by 32 threads: 32t * 160c * 1MB = 5 GB RAM. So it is not clear why CH needs another 15 GB to execute this query. 

Is there a way to use a pipeline like the following one?

1. Read only columns that appear in WHERE and ORDER BY clauses from disk;
2. Mark locations of each row that satisfy WHERE clause, heap sort on-the-fly;
3. Extract marked rows after scan is finished.

This way I wouldn't need tens of GB of RAM to perform deep "historical" searches.

If that is not possible at the moment, are there any plans to introduce such a pipeline in a future releases?

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Reduce memory usage and number of read rows in queries with `ORDER BY primary_key`.

Fixes #20553
Fixes #16250.

### CODE REVIEW

File: src/Processors/Merges/IMergingTransform.cpp [hunk 1]
```diff
@@ -77,12 +76,17 @@ IProcessor::Status IMergingTransformBase::prepareInitializeInputs()
             continue;
         }
 
-        auto chunk = input.pull();
+        /// setNotNeeded after reading first chunk, because in optimismtic case
+        /// (e.g. with optimized 'ORDER BY primary_key LIMIT n' and small 'n')
+        /// we won't have to read any chunks anymore;
+        auto chunk = input.pull(expected_one_block);
```
Comment (CurtizJ) [inline src/Processors/Merges/IMergingTransform.cpp:82]: Actual improvement is here.

File: src/Storages/MergeTree/MergeTreeInOrderSelectProcessor.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    auto size_predictor = (preferred_block_size_bytes == 0)
+        ? nullptr
+        : std::make_unique<MergeTreeBlockSizePredictor>(data_part, ordered_names, metadata_snapshot->getSampleBlock());
+
+    MarkRanges mark_ranges_for_task;
+    if (one_range_per_task)
+    {
+        mark_ranges_for_task = { std::move(all_mark_ranges.front()) };
+        all_mark_ranges.pop_front();
+    }
```
Comment (CurtizJ) [inline src/Storages/MergeTree/MergeTreeInOrderSelectProcessor.cpp:58]: And here.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 3]
```diff
@@ -190,11 +191,15 @@ Pipe ReadFromMergeTree::readInOrder(
     bool use_uncompressed_cache)
 {
     Pipes pipes;
+    /// For reading in order it makes sense to read only
+    /// one range per task to reduce number of read rows.
+    bool one_range_per_task = read_type != ReadType::Default;
+
```
Comment (akuzm) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:197]: Is this relevant only for a case when we have a LIMIT and won't read all the data? Looks like it should make no difference for a full scan...

File: src/Processors/Merges/ReplacingSortedTransform.h [hunk 4]
```diff
@@ -18,7 +18,7 @@ class ReplacingSortedTransform final : public IMergingTransform<ReplacingSortedA
         WriteBuffer * out_row_sources_buf_ = nullptr,
         bool use_average_block_sizes = false)
         : IMergingTransform(
-            num_inputs, header, header, true,
+            num_inputs, header, header, true, false,
```
Comment (akuzm) [inline src/Processors/Merges/ReplacingSortedTransform.h:21]: C++ really needs named parameters.... we could use C++20 designated initializers and create a parameter struct, but it requires you to set every field, unlike C, so it's also ugly. I tend to write something like `/* expected_one_block = */ false`. clang-tidy even checks that these comment match the function declaration.

File: src/Processors/QueryPlan/FinishSortingStep.cpp [hunk 5]
```diff
@@ -58,11 +58,14 @@ void FinishSortingStep::transformPipeline(QueryPipeline & pipeline, const BuildQ
...
         UInt64 limit_for_merging = (need_finish_sorting ? 0 : limit);
+        bool expected_one_block = limit_for_merging && limit_for_merging < max_block_size;
         auto transform = std::make_shared<MergingSortedTransform>(
                 pipeline.getHeader(),
                 pipeline.getNumStreams(),
                 prefix_description,
-                max_block_size, limit_for_merging);
+                max_block_size,
+                limit_for_merging,
+                expected_one_block);
```
Comment (akuzm) [inline src/Processors/QueryPlan/FinishSortingStep.cpp:68]: So this is the place where this entire logic is switched on, right? `expected_one_block` is set, and `read_one_range` is also set, or how was it called... maybe we need to invent some unified name  for these flags, like `has_limit_below_one_block`.

**Review by akuzm — COMMENTED** [review]
Looks fine.

Do you mind also adding some performance test for this? It isn't shown in any reports, but we actually track the memory usage for all perf test queries, so it might be helpful to have a representative query.

**Comment by akuzm** [comment]
```
SELECT *
FROM query_metrics_v2
WHERE (new_sha = '2165ff679a5a800dc88827139d9665c8e3772043') AND (abs(diff) > stat_threshold) AND (metric = 'memory_usage') AND (diff > 0.05)

Ok.

0 rows in set. Elapsed: 0.051 sec. Processed 180.22 thousand rows, 13.51 MB (3.52 million rows/s., 263.59 MB/s.) 
```

Looks like we don't have any representative queries in the perf tests yet.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp (hunk 3 above)
Comment (CurtizJ) [reply src/Processors/QueryPlan/ReadFromMergeTree.cpp]: Yes. I've added a pushdown of limit.

**Comment by CurtizJ** [comment]
> Do you mind also adding some performance test for this

I think it's ok to have current tests, because optimization of memory consumption affects also number of read rows and the time of query in result. Also I've added a functional test to check `max_memory_usage`.

**Pull request merged**